### PR TITLE
New Condition classes [10316]

### PR DESCRIPTION
**Pull request by jparisu**

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3556/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/746/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3156/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3588/)

File: include/fastdds/dds/core/condition/Condition.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ */
+class Condition
+{
+public:
+
+    // Condition class not implemented.
+
+    /**
+     * @brief Retrieves the trigger_value
```
Comment (rsanchez15) [inline include/fastdds/dds/core/condition/Condition.hpp:42]:
```suggestion
     * @brief Retrieves the trigger_value of the Condition
```

File: include/fastdds/dds/core/condition/Condition.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#include <fastrtps/fastrtps_dll.h>
+#include <vector>
+#include <fastdds/dds/log/Log.hpp>
+
+namespace eprosima {
+namespace fastdds {
+namespace dds {
+
+/**
+ * @brief The Condition class is the root base class  for all the conditions that may be attached to a WaitSet.
```
Comment (rsanchez15) [inline include/fastdds/dds/core/condition/Condition.hpp:32]:
```suggestion
 * @brief The Condition class is the root base class for all the conditions that may be attached to a WaitSet.
```

File: include/fastdds/dds/core/condition/Condition.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ *
```
Comment (rsanchez15) [inline include/fastdds/dds/core/condition/Condition.hpp:33]: In general
```suggestion
```

File: include/fastdds/dds/core/condition/Condition.hpp (hunk 1 above)
Comment (rsanchez15) [reply include/fastdds/dds/core/condition/Condition.hpp:42]: Also reference `trigger_value`

File: include/fastdds/dds/core/condition/Condition.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ */
+class Condition
+{
+public:
+
+    // Condition class not implemented.
+
+    /**
+     * @brief Retrieves the trigger_value
+     * @return true if trigger_value is set to true, false otherwise
```
Comment (rsanchez15) [inline include/fastdds/dds/core/condition/Condition.hpp:43]: In general, user typewriter font to refer a word of code

File: include/fastdds/dds/core/condition/GuardCondition.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#include <fastrtps/types/TypesBase.h>
+
+namespace eprosima {
+namespace fastdds {
+namespace dds {
+
+/**
+ * @brief The GuardCondition class is a specific Condition whose trigger_value is completely under the control
+ * of the application.
+ *
```
Comment (rsanchez15) [inline include/fastdds/dds/core/condition/GuardCondition.hpp:34]: Add also this

The purpose of the GuardCondition is to provide the means for the application to manually wakeup a WaitSet. This is
accomplished by attaching the GuardCondition to the WaitSet and then setting the trigger_value by means of the
set_trigger_value operation.

**Review by IkerLuengo — DISMISSED** [review]
LGTM

File: include/fastdds/dds/core/condition/GuardCondition.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+public:
+
+    // GuardCondition not implemented.
+
+    /**
+     * @brief Set the trigger_value
+     * @param value new value for trigger
+     * @return RETURN_OK
+     */
+    RTPS_DllAPI eprosima::fastrtps::types::ReturnCode_t set_trigger_value(
```
Comment (rsanchez15) [inline include/fastdds/dds/core/condition/GuardCondition.hpp:47]: In general, in all the other files in the library that have the namespace `eprosima::fastdds::dds`, ReturnCode_t it is declared as `using`.

File: test/unittest/dds/condition/ConditionTests.cpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
+// Copyright 2020 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (rsanchez15) [inline test/unittest/dds/condition/ConditionTests.cpp:1]:
```suggestion
// Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

File: test/unittest/dds/condition/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,45 @@
+# Copyright 2020 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (rsanchez15) [inline test/unittest/dds/condition/CMakeLists.txt:1]:
```suggestion
# Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

File: src/cpp/fastdds/core/condition/StatusCondition.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+fastrtps::types::ReturnCode_t StatusCondition::set_enabled_statuses(
+        const StatusMask& mask)
+{
+    static_cast<void>(mask);
+    return fastrtps::types::ReturnCode_t::RETCODE_UNSUPPORTED;
+}
+
+const StatusMask& StatusCondition::get_enabled_statuses() const
+{
+    logWarning(CONDITION, "get_enabled_statuses method not implemented");
```
Comment (rsanchez15) [inline src/cpp/fastdds/core/condition/StatusCondition.cpp:36]: In general
```suggestion
    logWarning(CONDITION, "get_enabled_statuses public member function not implemented");
```

**Review by rsanchez15 — CHANGES_REQUESTED** [review]
Review the classes, functions and variables descriptions. There is missing information that can be found in the DDS standard. Also add references when mentioning other classes or functions, and use typewrite font when using code words.

File: include/fastdds/dds/core/condition/Condition.hpp (hunk 1 above)
Comment (jparisu) [reply include/fastdds/dds/core/condition/Condition.hpp:42]: :+1:

File: include/fastdds/dds/core/condition/Condition.hpp (hunk 2 above)
Comment (jparisu) [reply include/fastdds/dds/core/condition/Condition.hpp:32]: :+1:

File: include/fastdds/dds/core/condition/Condition.hpp (hunk 3 above)
Comment (jparisu) [reply include/fastdds/dds/core/condition/Condition.hpp:33]: :+1:

File: include/fastdds/dds/core/condition/GuardCondition.hpp (hunk 5 above)
Comment (jparisu) [reply include/fastdds/dds/core/condition/GuardCondition.hpp:34]: :+1:

File: include/fastdds/dds/core/condition/Condition.hpp (hunk 4 above)
Comment (jparisu) [reply include/fastdds/dds/core/condition/Condition.hpp:43]: :+1:

File: test/unittest/dds/condition/ConditionTests.cpp (hunk 7 above)
Comment (jparisu) [reply test/unittest/dds/condition/ConditionTests.cpp:1]: :+1:

File: include/fastdds/dds/core/condition/GuardCondition.hpp (hunk 6 above)
Comment (jparisu) [reply include/fastdds/dds/core/condition/GuardCondition.hpp:47]: :+1:

File: test/unittest/dds/condition/CMakeLists.txt (hunk 8 above)
Comment (jparisu) [reply test/unittest/dds/condition/CMakeLists.txt:1]: :+1:

File: src/cpp/fastdds/core/condition/StatusCondition.cpp (hunk 9 above)
Comment (jparisu) [reply src/cpp/fastdds/core/condition/StatusCondition.cpp:36]: :+1:

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3604/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/794/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3192/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3636/)

File: include/fastdds/dds/core/condition/Condition.hpp (hunk 4 above)
Comment (rsanchez15) [reply include/fastdds/dds/core/condition/Condition.hpp]: The doxygen formatting and descriptions will be refactored in other PR.

File: include/fastdds/dds/core/condition/GuardCondition.hpp (hunk 6 above)
Comment (rsanchez15) [reply include/fastdds/dds/core/condition/GuardCondition.hpp]: This is what I meant in this comment:

https://github.com/eProsima/Fast-DDS/blob/09a7ddc94c79800c0f580cba3ee57fc5bb5eae4c/src/cpp/fastdds/domain/DomainParticipantImpl.hpp#L37

**Review by rsanchez15 — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3608/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/797/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3195/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3639/)

**Comment by richiware** [comment]
Rerun Linux job. Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3626/)

**Comment by richiware** [comment]
All jobs passed successfully. Merging...

**Comment by weigewansui** [comment]
Hey sir @richiware ,
I saw you merged this empty WaitSet class, I'm new to FastDDS but I want to use WaitSet and StatusCondition. Since we have WaitSet in dds/core (not fastdds/dds/core), but include/dds directory is not installed. I was wondering how can I use the APIs defined in include/dds/core?
Thanks

**Comment by richiware** [comment]
Sorry for the delay answering. We were preparing our future WaitSet API. We've started recently the implementation of this feature and we hopes it will come with next version 2.4 or 2.5.

**Comment by weigewansui** [comment]
@richiware  Thank you very much sir

**Pull request merged**